### ISSUE TITLE
Process hangs after run() when allowUndefinedFacts=true

### ISSUE THREAD

**Issue by egzplicit** [CacheControl/json-rules-engine#96]
When an engine is created like so:

```
const engine = new Engine(rules, {
  allowUndefinedFacts: true
});
```

and the rules set contains 100 rules, each containing around 6 conditions (in an `all` group) and the engine is ran against 50 items like so:

```
const results = [];
for (let f=0; f < items.length; f++) {
    const result = await engine.run({ ...items[f] });
    results.push(result);
}
```

with each `item` not having a few of the conditions defined in the rules, the engine produces the results but then the node process hangs for a long time.

I tried profiling it and found this:

![profiling](https://i.imgur.com/QkolSQF.png)

I'm not sure what is going on, the `almanac::factValue()` returns a rejected Promise for an undefined fact:

https://github.com/CacheControl/json-rules-engine/blob/1583cd6ff8deaa2028cbad295101aeab907be2a8/src/almanac.js#L89-L91

and the `rule::evaluate()` method does catch it and handles it if `allowUndefinedFacts` is true.

https://github.com/CacheControl/json-rules-engine/blob/1583cd6ff8deaa2028cbad295101aeab907be2a8/src/rule.js#L165-L168

However, something is not working as expected and the process hangs for a long time before recovering. CPU usage also goes up to 100% (see the profiler screenshot).

As a quick hack I found that if instead of rejecting the Promise in the almanac I resolve it with a dummy value, it never hangs and the engine works as expected.

**Comment by egzplicit** [comment]
Also ignoring the process hanging, it seems `Promise.resolve(undefined)` in the almanac here:

https://github.com/CacheControl/json-rules-engine/blob/1583cd6ff8deaa2028cbad295101aeab907be2a8/src/almanac.js#L89-L91

is much faster than rejecting. Running the engine with a list of 110 rules (6 facts in each) against 168 items with the Promise.reject() takes around `2 seconds` (and then the process hangs for a while, after the engine finished running 168 times) compared to just `617ms `for the same rules & inputs when I have a `Promise.resolve(undefined)` instead of the rejection (it's obviously a hack but for testing reasons I gave it a go).

**Comment by leoDreamer** [comment]
@CacheControl this is a useful issue, may lead to memory leak,

**Comment by CacheControl** [comment]
@egzplicit I've just published **v3.0.0** which solves this issue.

Let me know if things haven't improved on your end.
